yarn-project/pxe/src/storage/private_event_data_provider/private_event_data_provider.ts

```diff
@@ -0,0 +1,137 @@
+import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
+import type { AztecAsyncArray, AztecAsyncKVStore, AztecAsyncMap } from '@aztec/kv-store';
+import type { EventSelector } from '@aztec/stdlib/abi';
+import type { AztecAddress } from '@aztec/stdlib/aztec-address';
+
+import type { DataProvider } from '../data_provider.js';
+
+interface PrivateEventEntry {
+  logContent: Buffer;
+  blockNumber: number;
+}
+
+/**
+ * Stores decrypted private event logs.
+ */
+export class PrivateEventDataProvider implements DataProvider {
+  #store: AztecAsyncKVStore;
+  /** Array storing the actual private event log entries containing the log content and block number */
+  #eventLogs: AztecAsyncArray<PrivateEventEntry>;
+  /** Map from contract_address_recipient_eventSelector to array of indices into #eventLogs for efficient lookup */
+  #eventLogIndex: AztecAsyncMap<string, number[]>;
+  /**
+   * Map from tag to array of event log hashes.
+   * @dev A single tag can map to multiple hashes because we don't have a guarantee that tags won't be reused.
+   */
+  #tagToLogHashes: AztecAsyncMap<string, Buffer[]>;
+
+  logger = createLogger('private_event_data_provider');
+
+  constructor(store: AztecAsyncKVStore) {
+    this.#store = store;
+    this.#eventLogs = this.#store.openArray('private_event_logs');
+    this.#eventLogIndex = this.#store.openMap('private_event_log_index');
+    this.#tagToLogHashes = this.#store.openMap('tag_to_log_hashes');
+  }
+
+  /**
+   * Store a private event log.
+   * @param tag - The tag of the event log.
+   * @param contractAddress - The address of the contract that emitted the event.
+   * @param recipient - The recipient of the event.
+   * @param eventSelector - The event selector of the event.
+   * @param logContent - The content of the event.
+   * @param blockNumber - The block number in which the event was emitted.
+   */
+  storePrivateEventLog(
+    tag: Fr,
+    contractAddress: AztecAddress,
+    recipient: AztecAddress,
+    eventSelector: EventSelector,
+    logContent: Fr[],
+    blockNumber: number,
+  ): Promise<void> {
+    return this.#store.transactionAsync(async () => {
+      const key = `${contractAddress.toString()}_${recipient.toString()}_${eventSelector.toString()}`;
+
+      // Since we don't have a guarantee that the event log will be processed only once, we need to check whether
+      // the log has already been stored under the same tag to not have duplicate entries. It could also happen that
+      // there is a different log under the same tag, so we need to check for that as well.
+      const tagStr = tag.toString();
+      const logHash = sha256(serializeToBuffer([eventSelector.toField(), ...logContent]));
+
+      // Check if events with this tag already exist
+      const existingHashes = (await this.#tagToLogHashes.getAsync(tagStr)) || [];
+      const isDuplicate = existingHashes.some(hash => hash.equals(logHash));
```

Comment:
This is not sound. The same tag can exist multiple times, and two different events can have the exact same content.

The way to actually identify a log is by its tx hash and index in the log array for that tx. We can get this information from the `txEffects` call we're already making when processing notes.

This also makes me thing that 'log content' is a bad name, as here I thought it was the content of the log (i.e. the full tag plus ciphertext). Keep the name here, but we need to rename that concept in our log encoding scheme.

Reply:
Incidentally this likely means we no longer need the tag.

Reply:
You are right. Stupid Jans

Fixed in https://github.com/AztecProtocol/aztec-packages/pull/12957/commits/ee86781a60bdd3274f1a16b173f992c1e3ba4a88